--- a/net/minecraft/world/entity/monster/Strider.java
+++ b/net/minecraft/world/entity/monster/Strider.java
@@ -277,7 +_,13 @@
          BlockState blockstate = this.f_19853_.m_8055_(this.m_20183_());
          BlockState blockstate1 = this.m_217002_();
          boolean flag = blockstate.m_204336_(BlockTags.f_13086_) || blockstate1.m_204336_(BlockTags.f_13086_) || this.m_204036_(FluidTags.f_13132_) > 0.0D;
-         this.m_33951_(!flag);
+         // CraftBukkit start
+         if (!flag ^ this.m_33935_()) {
+            if (org.bukkit.craftbukkit.v1_19_R1.event.CraftEventFactory.callStriderTemperatureChangeEvent(this, !flag)) {
+               this.m_33951_(!flag);
+            }
+         }
+         // CraftBukkit end
       }
 
       super.m_8119_();
@@ -418,6 +_,7 @@
 
    private SpawnGroupData m_33881_(ServerLevelAccessor p_33882_, DifficultyInstance p_33883_, Mob p_33884_, @Nullable SpawnGroupData p_33885_) {
       p_33884_.m_7678_(this.m_20185_(), this.m_20186_(), this.m_20189_(), this.m_146908_(), 0.0F);
+      if (!net.minecraftforge.event.ForgeEventFactory.doSpecialSpawn(p_33884_, p_33882_, (float)p_33884_.m_20185_(), (float)p_33884_.m_20186_(), (float)p_33884_.m_20189_(), null, MobSpawnType.JOCKEY))
       p_33884_.m_6518_(p_33882_, p_33883_, MobSpawnType.JOCKEY, p_33885_, (CompoundTag)null);
       p_33884_.m_7998_(this, true);
       return new AgeableMob.AgeableMobGroupData(0.0F);
