--- a/net/minecraft/client/renderer/chunk/RenderChunkRegion.java
+++ b/net/minecraft/client/renderer/chunk/RenderChunkRegion.java
@@ -44,6 +_,11 @@
       return this.f_112908_.m_7717_(p_112940_, p_112941_);
    }
 
+   @Override
+   public net.minecraftforge.client.model.data.ModelDataManager getModelDataManager() {
+      return f_112908_.getModelDataManager();
+   }
+
    public LevelLightEngine m_5518_() {
       return this.f_112908_.m_5518_();
    }
@@ -65,5 +_,10 @@
 
    public int m_141928_() {
       return this.f_112908_.m_141928_();
+   }
+
+   @Override
+   public float getShade(float normalX, float normalY, float normalZ, boolean shade) {
+      return this.f_112908_.getShade(normalX, normalY, normalZ, shade);
    }
 }
