--- a/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
+++ b/net/minecraft/world/level/block/grower/AbstractTreeGrower.java
@@ -3,24 +_,37 @@
 import javax.annotation.Nullable;
 import net.minecraft.core.BlockPos;
 import net.minecraft.core.Holder;
+import net.minecraft.data.worldgen.features.TreeFeatures;
 import net.minecraft.server.level.ServerLevel;
 import net.minecraft.tags.BlockTags;
 import net.minecraft.util.RandomSource;
 import net.minecraft.world.level.LevelAccessor;
+import net.minecraft.world.level.block.SaplingBlock;
 import net.minecraft.world.level.block.state.BlockState;
 import net.minecraft.world.level.chunk.ChunkGenerator;
 import net.minecraft.world.level.levelgen.feature.ConfiguredFeature;
+import org.bukkit.TreeType;
 
 public abstract class AbstractTreeGrower {
+   // Forge: context-sensitive version to be able to use data pack provided features
+   @Nullable
+   protected Holder<? extends ConfiguredFeature<?, ?>> getConfiguredFeature(ServerLevel level, ChunkGenerator chunkGenerator, BlockPos pos, BlockState state, RandomSource random, boolean hasFlowers) {
+      return m_213888_(random, hasFlowers);
+   }
+
+   /** @deprecated Forge: Use {@link #getConfiguredFeature(ServerLevel, ChunkGenerator, BlockPos, BlockState, RandomSource, boolean)} instead */
+   @Deprecated
    @Nullable
    protected abstract Holder<? extends ConfiguredFeature<?, ?>> m_213888_(RandomSource p_222910_, boolean p_222911_);
 
    public boolean m_213817_(ServerLevel p_222905_, ChunkGenerator p_222906_, BlockPos p_222907_, BlockState p_222908_, RandomSource p_222909_) {
-      Holder<? extends ConfiguredFeature<?, ?>> holder = this.m_213888_(p_222909_, this.m_60011_(p_222905_, p_222907_));
-      if (holder == null) {
+      Holder<? extends ConfiguredFeature<?, ?>> holder = this.getConfiguredFeature(p_222905_, p_222906_, p_222907_, p_222908_, p_222909_, this.m_60011_(p_222905_, p_222907_));
+      net.minecraftforge.event.level.SaplingGrowTreeEvent event = net.minecraftforge.event.ForgeEventFactory.blockGrowFeature(p_222905_, p_222909_, p_222907_, holder);
+      if (event.getResult().equals(net.minecraftforge.eventbus.api.Event.Result.DENY) || event.getFeature() == null) {
          return false;
       } else {
-         ConfiguredFeature<?, ?> configuredfeature = holder.m_203334_();
+         setTreeType(holder); // CraftBukkit
+         ConfiguredFeature<?, ?> configuredfeature = event.getFeature().m_203334_();
          BlockState blockstate = p_222905_.m_6425_(p_222907_).m_76188_();
          p_222905_.m_7731_(p_222907_, blockstate, 4);
          if (configuredfeature.m_224953_(p_222905_, p_222906_, p_222909_, p_222907_)) {
@@ -35,6 +_,55 @@
          }
       }
    }
+
+   // CraftBukkit start
+   protected void setTreeType(Holder<? extends ConfiguredFeature<?, ?>> worldgentreeabstract) {
+      if (worldgentreeabstract == TreeFeatures.f_195123_ || worldgentreeabstract == TreeFeatures.f_195142_) {
+         SaplingBlock.treeType = TreeType.TREE;
+      } else if (worldgentreeabstract == TreeFeatures.f_195122_) {
+         SaplingBlock.treeType = TreeType.RED_MUSHROOM;
+      } else if (worldgentreeabstract == TreeFeatures.f_195121_) {
+         SaplingBlock.treeType = TreeType.BROWN_MUSHROOM;
+      } else if (worldgentreeabstract == TreeFeatures.f_195129_) {
+         SaplingBlock.treeType = TreeType.COCOA_TREE;
+      } else if (worldgentreeabstract == TreeFeatures.f_195131_) {
+         SaplingBlock.treeType = TreeType.SMALL_JUNGLE;
+      } else if (worldgentreeabstract == TreeFeatures.f_195128_) {
+         SaplingBlock.treeType = TreeType.TALL_REDWOOD;
+      } else if (worldgentreeabstract == TreeFeatures.f_195127_) {
+         SaplingBlock.treeType = TreeType.REDWOOD;
+      } else if (worldgentreeabstract == TreeFeatures.f_195126_) {
+         SaplingBlock.treeType = TreeType.ACACIA;
+      } else if (worldgentreeabstract == TreeFeatures.f_195125_ || worldgentreeabstract == TreeFeatures.f_195108_) {
+         SaplingBlock.treeType = TreeType.BIRCH;
+      } else if (worldgentreeabstract == TreeFeatures.f_195135_) {
+         SaplingBlock.treeType = TreeType.TALL_BIRCH;
+      } else if (worldgentreeabstract == TreeFeatures.f_195137_) {
+         SaplingBlock.treeType = TreeType.SWAMP;
+      } else if (worldgentreeabstract == TreeFeatures.f_195130_ || worldgentreeabstract == TreeFeatures.f_195111_) {
+         SaplingBlock.treeType = TreeType.BIG_TREE;
+      } else if (worldgentreeabstract == TreeFeatures.f_195138_) {
+         SaplingBlock.treeType = TreeType.JUNGLE_BUSH;
+      } else if (worldgentreeabstract == TreeFeatures.f_195124_) {
+         SaplingBlock.treeType = TreeType.DARK_OAK;
+      } else if (worldgentreeabstract == TreeFeatures.f_195133_) {
+         SaplingBlock.treeType = TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == TreeFeatures.f_195134_) {
+         SaplingBlock.treeType = TreeType.MEGA_REDWOOD;
+      } else if (worldgentreeabstract == TreeFeatures.f_195132_) {
+         SaplingBlock.treeType = TreeType.JUNGLE;
+      } else if (worldgentreeabstract == TreeFeatures.f_195139_) {
+         SaplingBlock.treeType = TreeType.AZALEA;
+      } else if (worldgentreeabstract == TreeFeatures.f_236762_) {
+         SaplingBlock.treeType = TreeType.MANGROVE;
+      } else if (worldgentreeabstract == TreeFeatures.f_236763_) {
+         SaplingBlock.treeType = TreeType.TALL_MANGROVE;
+      } else {
+         // LoliServer - handle MOD_CUSTOM treeType
+         SaplingBlock.treeType = TreeType.MOD_CUSTOM;
+      }
+   }
+   // CraftBukkit end
 
    private boolean m_60011_(LevelAccessor p_60012_, BlockPos p_60013_) {
       for(BlockPos blockpos : BlockPos.MutableBlockPos.m_121940_(p_60013_.m_7495_().m_122013_(2).m_122025_(2), p_60013_.m_7494_().m_122020_(2).m_122030_(2))) {
