--- a/net/minecraft/world/entity/decoration/ItemFrame.java
+++ b/net/minecraft/world/entity/decoration/ItemFrame.java
@@ -68,7 +_,7 @@
       this.m_20088_().m_135372_(f_31758_, 0);
    }
 
-   protected void m_6022_(Direction p_31793_) {
+   public void m_6022_(Direction p_31793_) {
       Validate.notNull(p_31793_);
       this.f_31699_ = p_31793_;
       if (p_31793_.m_122434_().m_122479_()) {
@@ -86,15 +_,23 @@
 
    protected void m_7087_() {
       if (this.f_31699_ != null) {
+         // CraftBukkit start code moved in to calculateBoundingBox
+         this.m_20011_(calculateBoundingBox(this, this.f_31698_, this.f_31699_, this.m_7076_(), this.m_7068_()));
+         // CraftBukkit end
+      }
+   }
+
+   /*
+   protected void recalculateBoundingBox() {
          double d0 = 0.46875D;
-         double d1 = (double)this.f_31698_.m_123341_() + 0.5D - (double)this.f_31699_.m_122429_() * 0.46875D;
-         double d2 = (double)this.f_31698_.m_123342_() + 0.5D - (double)this.f_31699_.m_122430_() * 0.46875D;
-         double d3 = (double)this.f_31698_.m_123343_() + 0.5D - (double)this.f_31699_.m_122431_() * 0.46875D;
-         this.m_20343_(d1, d2, d3);
-         double d4 = (double)this.m_7076_();
-         double d5 = (double)this.m_7068_();
-         double d6 = (double)this.m_7076_();
-         Direction.Axis direction$axis = this.f_31699_.m_122434_();
+         double d1 = (double)this.pos.getX() + 0.5D - (double)this.direction.getStepX() * 0.46875D;
+         double d2 = (double)this.pos.getY() + 0.5D - (double)this.direction.getStepY() * 0.46875D;
+         double d3 = (double)this.pos.getZ() + 0.5D - (double)this.direction.getStepZ() * 0.46875D;
+         this.setPosRaw(d1, d2, d3);
+         double d4 = (double)this.getWidth();
+         double d5 = (double)this.getHeight();
+         double d6 = (double)this.getWidth();
+         Direction.Axis direction$axis = this.direction.getAxis();
          switch (direction$axis) {
             case X:
                d4 = 1.0D;
@@ -109,8 +_,38 @@
          d4 /= 32.0D;
          d5 /= 32.0D;
          d6 /= 32.0D;
-         this.m_20011_(new AABB(d1 - d4, d2 - d5, d3 - d6, d1 + d4, d2 + d5, d3 + d6));
-      }
+         this.setBoundingBox(new AABB(d1 - d4, d2 - d5, d3 - d6, d1 + d4, d2 + d5, d3 + d6));
+      }
+    */
+
+   // CraftBukkit start - break out BB calc into own method
+   public static AABB calculateBoundingBox(@Nullable Entity entity, BlockPos blockPosition, Direction direction, int width, int height) {
+      double d0 = 0.46875D;
+      double d1 = (double) blockPosition.m_123341_() + 0.5D - (double) direction.m_122429_() * 0.46875D;
+      double d2 = (double) blockPosition.m_123342_() + 0.5D - (double) direction.m_122430_() * 0.46875D;
+      double d3 = (double) blockPosition.m_123343_() + 0.5D - (double) direction.m_122431_() * 0.46875D;
+      if (entity != null) {
+         entity.m_20343_(d1, d2, d3);
+      }
+      double d4 = (double) width;
+      double d5 = (double) height;
+      double d6 = (double) width;
+      Direction.Axis direction$axis = direction.m_122434_();
+      switch (direction$axis) {
+         case X:
+            d4 = 1.0D;
+            break;
+         case Y:
+            d5 = 1.0D;
+            break;
+         case Z:
+            d6 = 1.0D;
+      }
+
+      d4 /= 32.0D;
+      d5 /= 32.0D;
+      d6 /= 32.0D;
+      return new AABB(d1 - d4, d2 - d5, d3 - d6, d1 + d4, d2 + d5, d3 + d6);
    }
 
    public boolean m_7088_() {
@@ -154,6 +_,11 @@
          return false;
       } else if (!p_31776_.m_19372_() && !this.m_31822_().m_41619_()) {
          if (!this.f_19853_.f_46443_) {
+            // CraftBukkit start - fire EntityDamageEvent
+            if (org.bukkit.craftbukkit.v1_19_R1.event.CraftEventFactory.handleNonLivingEntityDamageEvent(this, p_31776_, p_31777_, false) || this.m_213877_()) {
+               return true;
+            }
+            // CraftBukkit end
             this.m_31802_(p_31776_.m_7639_(), false);
             this.m_5496_(this.m_142544_(), 1.0F, 1.0F);
          }
@@ -278,6 +_,24 @@
       this.m_218865_(p_31790_);
       this.m_20088_().m_135381_(f_31757_, p_31790_);
       if (!p_31790_.m_41619_()) {
+         this.m_5496_(this.m_142546_(), 1.0F, 1.0F);
+      }
+
+      if (p_31791_ && this.f_31698_ != null) {
+         this.f_19853_.m_46717_(this.f_31698_, Blocks.f_50016_);
+      }
+
+   }
+
+   public void setItem(ItemStack p_31790_, boolean p_31791_, boolean playSound) {
+      if (!p_31790_.m_41619_()) {
+         p_31790_ = p_31790_.m_41777_();
+         p_31790_.m_41764_(1);
+         p_31790_.m_41636_(this);
+      }
+
+      this.m_20088_().m_135381_(f_31757_, p_31790_);
+      if (!p_31790_.m_41619_() && playSound) { // CraftBukkit
          this.m_5496_(this.m_142546_(), 1.0F, 1.0F);
       }
 
